### ISSUE TITLE
Exception when evaluating in repl after hot-reload of `cljs.user`

### ISSUE THREAD

**Issue by grav** [thheller/shadow-cljs#905]
Whenever I edit something in `cljs.user` ns, and shadow-cljs hot-reloads that particular namesppace, I get an exception when evaluating anything:

```
[2021-06-22 22:45:14.943 - WARNING] :shadow.cljs.devtools.server.worker.impl/cljs-compile-ex - {:input {:code "cljs.user/foo", :ns cljs.user, :repl true}}
ExceptionInfo Failed to process REPL command {:eof? false, :ns cljs.user, :form cljs.user/foo, :source "cljs.user/foo", :tag :shadow.cljs.repl/process-ex}
	shadow.cljs.repl/process-read-result (repl.clj:501)
	shadow.cljs.repl/process-read-result (repl.clj:475)
	shadow.cljs.repl/process-input (repl.clj:659)
	shadow.cljs.repl/process-input (repl.clj:637)
	shadow.cljs.devtools.server.worker.impl/eval16193/fn--16196 (impl.clj:685)
	clojure.lang.MultiFn.invoke (MultiFn.java:234)
	shadow.cljs.devtools.server.util/server-thread/fn--15837/fn--15838/fn--15846 (util.clj:284)
	shadow.cljs.devtools.server.util/server-thread/fn--15837/fn--15838 (util.clj:283)
	shadow.cljs.devtools.server.util/server-thread/fn--15837 (util.clj:256)
	java.lang.Thread.run (Thread.java:829)
Caused by:
ExceptionInfo no source by provide: cljs.user {:provide cljs.user}
	shadow.build.data/get-source-id-by-provide (data.clj:186)
	shadow.build.data/get-source-id-by-provide (data.clj:183)
	shadow.build.data/get-source-by-provide (data.clj:189)
	shadow.build.data/get-source-by-provide (data.clj:188)
	shadow.cljs.repl/repl-compile/fn--15628/fn--15629 (repl.clj:437)
	shadow.cljs.repl/repl-compile/fn--15628 (repl.clj:413)
	shadow.cljs.repl/repl-compile (repl.clj:411)
	shadow.cljs.repl/repl-compile (repl.clj:408)
	shadow.cljs.repl/process-read-result (repl.clj:499)
	shadow.cljs.repl/process-read-result (repl.clj:475)
```

I've create a minimal repro here: https://github.com/grav/shadow-cljs-user-issue

Also briefly discussed issue here:
https://clojurians.slack.com/archives/C6N245JGG/p1624392092223000

**Comment by thheller** [comment]
The Slack history unfortunately is gone. Can't remember what exactly we discussed regarding this.

The problem is that `cljs.user` is not part of the regular build and is added dynamically after the fact when the REPL connects. Normally those dynamically added namespaces are not supposed to be part of the hot-reload cycle at all. So the bug here is that it even attempts to hot-reload this in the first place. It will remove sources that aren't part of the build, which leads to the "no source" error since it was removed but the hot-reload code is confused and still tries to access it.

One way to make it part of the build is using `:preloads` so changing the `:main` module to `{:init-fn foo.bar/main :preloads [cljs.user]}` fixes it.

Is the expectation here that `cljs.user` should hot-reload in the first place? My take on this was that everything that happens in the REPL should not play into the hot-reload cycle at all as you can easily craft circular dependencies and such that way. The bug here is basically that `cljs/user.cljs` exists so the code treats it as a regular file. Normally `cljs.user` doesn't have an actual file and other things defined in the REPL via `ns` also don't have files.

**Comment by lilactown** [comment]
This also occurs if one require's something from a REPL, i.e. `(require '[some.other.ns])` and then modify `some.other.ns` on disk. I lose the ability to refer to any vars in `some.other.ns` and I cannot `require` it again.

My solution for now is to avoid writing `some.other.ns` to disk until I'm done modifying it, continuing to evaluate it at the REPL, and once I do save it, I restart the build watch. At least I can do that without shutting shadow-cljs down:

```
cljs.user=> :cljs/quit
shadow.user=> (shadow/stop-worker :app)
shadow.user=> (shadow/watch :app)
shadow.user=> (shadow/repl :app)
cljs.user=> 
```

and reload the page

**Comment by thheller** [comment]
@lilactown I cannot reproduce your description. If the ns is on disk and not part of the regular build it is not hot-reloaded at all?

**Comment by grav** [comment]
> My take on this was that everything that happens in the REPL should not play into the hot-reload cycle at all as you can easily craft circular dependencies and such that way

@thheller Agree! Thanks for the fix!

**Comment by lilactown** [comment]
> @lilactown I cannot reproduce your description. If the ns is on disk and not part of the regular build it is not hot-reloaded at all?

It would trigger a hot reload of the build, but I could no longer refer to the ns on disk that had been only required via the REPL.

**Comment by thheller** [comment]
You are still speaking in riddles :P 

> but I could no longer refer to the ns on disk

What does this mean? If that means you had it required and aliased in some ns which is actually part of the build then that alias will be gone if that ns hot-reloaded. That is working as intended.

eg.

```
(require 'my.app)
(in-ns 'my.app)
(require '[something.else :as x])
x/foo
;; now editing my.app and it gets hot-reloaded
x/foo
;; fails since the x/ alias is gone since the file on disk doesn't have it and hot-reload only accounts for disk
```

**Comment by lilactown** [comment]
I mean that I have a namespace on disk, part of the classpath, but that is not a part of the build, i.e. is not a part of the dependency tree of any part of my app:

```clojure
(ns just.dev.things)

(def foo :foo)
```

I then require it from the REPL:

```
cljs.user> (require '[just.dev.things :as jdt])
nil
cljs.user> jdt/foo
:foo
```

I can modify my application code and it will hot reload fine. I can still reference `jdt/foo` just fine.

---

However, if I edit `just.dev.things` and save it to disk...

```clojure
(ns just.dev.things)

(def foo :bar)
```

This triggers a hot reload (wrong) and also, I can no longer reference it from my REPL:

```
cljs.user> jdt/foo
------ WARNING - :undeclared-ns ------------------------------------------------
 Resource: <eval>:1:1
 No such namespace: jdt, could not locate jdt.cljs, jdt.cljc, or JavaScript source providing "jdt"
--------------------------------------------------------------------------------

------ WARNING - :undeclared-var -----------------------------------------------
 Resource: <eval>:1:1
 Use of undeclared Var jdt/foo
--------------------------------------------------------------------------------

nil
cljs.user> just.dev.things/foo
------ WARNING - :undeclared-var -----------------------------------------------
 Resource: <eval>:1:1
 Use of undeclared Var just.dev.things/foo
--------------------------------------------------------------------------------

nil
```

Attempting to require it again from the REPL, results in the REPL timing out:

```
cljs.user> (require '[just.dev.things :as jdt])
Timeout while waiting for result.
```

Otherwise, REPL invocations work fine, but I can no longer refer to `just.dev.things` from the REPL. To recover I have to restart the watch.

**Comment by lilactown** [comment]
@thheller this issue seems to be fixed in 2.15.6. In the above example, modifying `just.dev.things` does not trigger a hot reload and I am able to continue to refer to it after saving it to disk. To see changes I can evaluate forms from my editor. 👍🏻
